p4-16/spec/P4-16-spec.mdk

```diff
@@ -6007,6 +6007,19 @@ Figure [#fig-subparser] shows a diagram of this process.
 Note that since P4 requires declarations to precede uses, it is impossible to
 create recursive (or mutually recursive) parsers.
 
+When a parser is instantiated, all its local declarations of stateful instantiations are evaluated recursively.
+Each instantiation of a parser pertains to the new instance a unique set of value sets, extern objects, and inner parser instances.
+Thus, invoking a parser instance twice is different from invoking two parser instances each once, where the former
+accesses the same stateful constructs while the latter access two different copies.
+
+This exactly-once evaluation only applies to stateful instantiations. For local variable declarations,
+whether in `state`s or out, and whether with initializers or not, they are always evaluated
+when a parser instance is invoked. That is, local variables in a parser never persist across invocations. For variables
+declared outside `state`s, they are evaluated at the beginning of execution.
+
+When a parser with no constructor parameters is instantiated exactly once, 
+it is possible to invoke the parser directly (see Section [#sec-direct-invocation]).
```

Comment:
Suggested alternate wording for the last paragraph: "The above applies to all parser instantiations, regardless of whether they are explicit instantiations, or performed via direct type invocation (see Section [#sec-direct-invocation])."